gateway/transaction.go

```diff
@@ -9,51 +9,36 @@ import (
 	"strings"
 	"time"
 
+	"github.com/dontpanicdao/caigo/types"
 	"github.com/google/go-querystring/query"
 )
 
-/*
-	StarkNet transaction states
-*/
-const (
-	NOT_RECIEVED = TxStatus(iota)
-	REJECTED
-	RECEIVED
-	PENDING
-	ACCEPTED_ON_L2
-	ACCEPTED_ON_L1
-)
-
-var statuses = []string{"NOT_RECEIVED", "REJECTED", "RECEIVED", "PENDING", "ACCEPTED_ON_L2", "ACCEPTED_ON_L1"}
-
-type TxStatus int
-
-type TransactionStatus struct {
-	TxStatus        string `json:"tx_status"`
-	BlockHash       string `json:"block_hash"`
-	TxFailureReason struct {
-		ErrorMessage string `json:"error_message,omitempty"`
-	} `json:"tx_failure_reason,omitempty"`
+type StarknetTransaction struct {
+	TransactionIndex int         `json:"transaction_index"`
+	BlockNumber      int         `json:"block_number"`
+	Transaction      Transaction `json:"transaction"`
+	BlockHash        string      `json:"block_hash"`
+	Status           string      `json:"status"`
 }
 
-// Starknet transaction composition
 type Transaction struct {
+	TransactionHash    string   `json:"transaction_hash,omitempty"`
 	ContractAddress    string   `json:"contract_address"`
 	EntryPointSelector string   `json:"entry_point_selector"`
 	Calldata           []string `json:"calldata"`
 	Signature          []string `json:"signature"`
 	EntryPointType     string   `json:"entry_point_type,omitempty"`
-	TransactionHash    string   `json:"transaction_hash,omitempty"`
 	Type               string   `json:"type,omitempty"`
 	Nonce              string   `json:"nonce,omitempty"`
 }
 
-type StarknetTransaction struct {
-	TransactionIndex int         `json:"transaction_index"`
-	BlockNumber      int         `json:"block_number"`
-	Transaction      Transaction `json:"transaction"`
-	BlockHash        string      `json:"block_hash"`
-	Status           string      `json:"status"`
+func (t Transaction) Normalize() *types.Transaction {
```

Comment:
I understand Normalize on the 'Block' type as its a different payload depending on the provider, but I don't think I understand here. 

Can we use the types.Transaction and type.TransactionReceipt here and omit the call to Normalize?

Reply:
My thought was the base types should conform to the official spec at https://github.com/starkware-libs/starknet-specs/blob/master/api/starknet_api_openrpc.json#L919

They are a subset of the gateway fields right now. I think that if people start depending on members that aren't part of the spec it will cause more migration pain later, since they might need to be removed eventually.

So, if a caller wants to ensure backward compatibility, they should use the provider interface/types. If they want all the gateway fields, they can use the Gateway client

Reply:
Cool, that makes sense.